cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java

```diff
@@ -411,15 +419,15 @@ protected void processMessages(Iterator<Message> iterator)
               false)
           .collect(Collectors.toList());
 
-      if (programFinishTime.get() == -1L) {
-        long finishTime = findProgramFinishTime(message);
-        if (finishTime >= 0) {
+      if (completionDetails.get() == null) {
+        ProgramRunCompletionDetails completionDetails = findProgramFinishInfo(message);
+        if (completionDetails != null) {
           detectedProgramFinish = true;
           LOG.trace("Detected program {} finish time {} in topic {}",
-              programRunId, finishTime,
+              programRunId, completionDetails.getEndTimestamp(),
               topicId.getTopic());
         }
-        programFinishTime.compareAndSet(-1L, finishTime);
+        RuntimeClientService.this.completionDetails.compareAndSet(null, completionDetails);
```

Comment:
Please name variable different from field. It makes it confusing and may be mixed up in refactorings.

Reply:
Done.